pipeline/astrometry.py

```diff
@@ -0,0 +1,121 @@
+import sqlalchemy as sa
+
+from pipeline.parameters import Parameters
+
+from models.base import SmartSession
+from models.image import Image
+
+from pipeline.utils import get_image_cache
+
+
+class ParsAstrometry(Parameters):
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.cross_match_catalog = self.add_par(
+            'cross_match_catalog',
+            'Gaia',
+            str,
+            'Which catalog should be used for cross matching for astrometry. '
+        )
+        self.add_alias('catalog', 'cross_match_catalog')
+
+        self._enforce_no_new_attrs = True
+
+        self.override(kwargs)
+
+    def _get_process_name(self):
+        return 'astrometry'
+
+    def _get_upstream_process_names(self):
+        return ['preprocessor']
+
+
+class Astrometry:
+    def __init__(self, **kwargs):
+        self.pars = ParsAstrometry()
+        self.pars.update(kwargs)
+
+    def run(self, exposure_id, ccd_id, image_cache=None, prov_cache=None, session=None):
+        """
+        Extract sources and use their positions to calculate the astrometric solution.
+
+        Parameters
+        ----------
+        exposure_id : int
```

Comment:
I really would like to keep the exposure_id/ccd_id throughout the whole pipeline. The same exp/ccd combination does not uniquely identify which image is picked. This is true for astrometry and for all other pipeline stages. 
The way I think about it is that if you must specify which provenance you are using in the upstream, and only then you can pinpoint the right image from the exp/ccd ids. This can happen in one of the following ways:
1) you gave the image in the image cache, so it is already loaded. 
2) you gave the provenance in the provenance cache, so it can be used to load a specific image (this is the least likely to happen, practically). 
3) you just load the "most recent" provenance and that identifies which image is associated with exp/ccd ids in the database.

Reply:
But how are you going to handle images that don't have an associated exposure?  For instance, references download from a survey, or stacks?

Reply:
I think that would be handled by the same object but using a different function, or maybe just add a `coadd_id` that when you use it, it overrides the exposure_id/ccd_id. 
The reason I went with the exp/ccd ids is that it lets the user easily pick what data to process, without needing to know the ids of the intermediate products.

Reply:
I agree that you want the user to be able to pass exposure name and chip number.  But, that can be handled by the code without the architecture of calling every function being tied to exposure name and chip number.

As I understand it, the real goal is to be able to track when the same image starts the pipeline a second time.  This means that a given exposure and CCD will have to have a unique signature, but it doesn't *have* to be (exposure id, ccd id).  We could define an "ImageSource" class or some such where each object has a unique id.  Then, we write as many methods as we need for that ImageSource based on where images come from -- it might be exposure plus ccd, it might be a downloaded survey, it might be a coadd with a specific coadd provnance signature, etc.   Indeed, there can be multiple subclasses of ImageSource to cover the various possibilities (though there will need to be a single method that knows all of them, so other code has a single place to call and doesn't have to try all of them).  All of those methods take what's needed, and return the unique id.  That is the thing that the rest of the methods sling around, and they can go back to that class to get more information as necessary.

When the code is run, the user can specify exposure name and chip, and the code calls the right method in this class to get the unique id.  Later, part of the code identifies wants a reference from a downloaded survey, so it calls whatever code identifies that, and then it can use *that* unique id to pass to astrometry, photometry, etc.

This is more flexible, because it doesn't bake in a specific way of identifying an image to every other routine we call later.  It's also cleaner, because most of the subsequent methods probably don't care about exposure and chip, they just need a way of identifying the image.  A unique handle for that is going to be a better way to do it than the specific data for one type (albeit the most common) of way of originating an image.